internal/resolution/variablesources/bundle_uniqueness.go

```diff
@@ -4,15 +4,52 @@ import (
 	"context"
 	"fmt"
 
+	"k8s.io/apimachinery/pkg/util/sets"
+
 	"github.com/operator-framework/deppy/pkg/deppy"
 	"github.com/operator-framework/deppy/pkg/deppy/input"
-	"k8s.io/apimachinery/pkg/util/sets"
 
 	"github.com/operator-framework/operator-controller/internal/catalogmetadata"
 	olmvariables "github.com/operator-framework/operator-controller/internal/resolution/variables"
 )
 
-var _ input.VariableSource = &CRDUniquenessConstraintsVariableSource{}
+// MakeBundleUniquenessVariables produces variables that constrain
+// the solution to at most 1 bundle per package.
+// These variables guarantee that no two versions of
+// the same package are running at the same time.
+func MakeBundleUniquenessVariables(bundleVariables []*olmvariables.BundleVariable) []*olmvariables.BundleUniquenessVariable {
+	result := []*olmvariables.BundleUniquenessVariable{}
+
+	bundleIDs := sets.Set[deppy.Identifier]{}
+	packageOrder := []string{}
+	bundleOrder := map[string][]deppy.Identifier{}
+	for _, bundleVariable := range bundleVariables {
+		bundles := []*catalogmetadata.Bundle{bundleVariable.Bundle()}
+		bundles = append(bundles, bundleVariable.Dependencies()...)
```

Comment:
Nit: we know size of bundles will be `1+len(bundleVariable.Dependencies)`. Does it make sense to allocate at the right size first?

Alternately, do we really even _need_ a bundles slice? Seems like we could make the `for` loop's contents its own function, and then do something like:

```go
updateMaps := func(bundles ...*olmvariables.BundleVariable) {
	for _, bundle := range bundles {
		id := olmvariables.BundleVariableID(bundle)
	 	// get bundleID package and update map
 		packageName := bundle.Package

		if _, ok := bundleOrder[packageName]; !ok {
	 		packageOrder = append(packageOrder, packageName)
 		}

	 	if !bundleIDs.Has(id) {
 			bundleIDs.Insert(id)
	 		bundleOrder[packageName] = append(bundleOrder[packageName], id)
 		}
	}
}
updateMaps(bundleVariable.Bundle())
updateMaps(bundleVariable.Dependencies()...)

Reply:
> Nit: we know size of bundles will be `1+len(bundleVariable.Dependencies)`. Does it make sense to allocate at the right size first?

Oh, yeah! Let's allocate once!

I'll update to something like this shortly:

```go
bundles := make([]*catalogmetadata.Bundle, 0, 1+len(bundleVariable.Dependencies()))
bundles = append(bundles, bundleVariable.Bundle())
bundles = append(bundles, bundleVariable.Dependencies()...)
```

Reply:
Updated

Reply:
> Alternately, do we really even _need_ a bundles slice? Seems like we could make the `for` loop's contents its own function, and then do something like

I think there still be slices in the variadic arugments. They are just implicit. So I went with the single allocaiton option.